ndonnx/_utility.py

```diff
@@ -64,17 +58,22 @@ def _promote_with_none(*args: Array | npt.ArrayLike) -> list[Array | None]:
                 dtype = np.dtype(f"int{eager_value.dtype.itemsize * 16}")
             scalar._set(scalar.astype(dtypes.from_numpy_dtype(dtype)))
 
-    target_dtype = result_type(*[arr for arr in arr_or_none if arr is not None])
+    target_dtype = ndx.result_type(*[arr for arr in arr_or_none if arr is not None])
+
+    if target_dtype is None:
+        return None
 
-    return [None if arr is None else astype(arr, target_dtype) for arr in arr_or_none]
+    return [None if arr is None else arr.astype(target_dtype) for arr in arr_or_none]
 
 
-def promote(*args: Array | npt.ArrayLike) -> list[Array]:
+def promote(*args: Array | npt.ArrayLike) -> list[Array] | None:
```

Comment:
This signature makes it very annoying to work with from a typing point of view. Not having `None` in the return type is the exact reason why we have `promote` and `_promote_with_none`.

Reply:
I think this is an aesthetic decision. The other options are an explicit result class or driving control flow through exception handling. I'm not sure either of those solutions is super pleasant.

Reply:
But it will nag us for every promotion between built-in data types. `None` is not very descriptive either. I think the most Pythonic thing would indeed be an exception.

Reply:
> I think the most Pythonic thing would indeed be an exception.

This PR was originally structured with this. However, it requires exposing a new exception type (there was a `PromotionError` since it is not safe to drive control-flow from an arbitrary `TypeError` down the stack. It also makes the logic in [`_binary`](https://github.com/Quantco/ndonnx/pull/17/files#diff-cd5d4a98029b4c5c571094c69a158fae5d5ee688a5e085614c13f68230984555R386) quite hacky with try-catches wrapping the function calls.

> But it will nag us for every promotion between built-in data types.

I agree that the ergonomics are unfortunate and I was hoping for something neater but I would really like to not treat built-in dtypes as anything special. You should be handling failure anyway if you're using `result_type` if the intention is to write safe code - type-promotion does not need to succeed always. Happy to think about it some more.

Reply:
Indeed, a custom exception would probably be good to have here. Maybe it is ok to expose that exception publicly. NumPy has [`DTypePromotionError`](https://numpy.org/doc/stable/reference/generated/numpy.exceptions.DTypePromotionError.html#numpy.exceptions.DTypePromotionError) in its public interface, too.